packages/cactus-plugin-ccmodel-hephaestus/src/main/typescript/plugin-ccmodel-hephaestus.ts

```diff
@@ -0,0 +1,911 @@
+import { Server } from "http";
+import { Server as SecureServer } from "https";
+import { Optional } from "typescript-optional";
+import { promisify } from "util";
+import {
+  IPluginWebService,
+  IWebServiceEndpoint,
+  ICactusPlugin,
+  ICactusPluginOptions,
+  LedgerType,
+} from "@hyperledger/cactus-core-api";
+
+import { RuntimeError } from "run-time-error-cjs";
+
+import fs from "fs";
+import path from "path";
+
+import { PluginRegistry } from "@hyperledger/cactus-core";
+import { Express } from "express";
+
+import {
+  Checks,
+  Logger,
+  LoggerProvider,
+  LogLevelDesc,
+} from "@hyperledger/cactus-common";
+import {
+  calculateGasPriceEth,
+  CarbonFootPrintConstants,
+} from "./models/carbon-footprint";
+import {
+  CrossChainEvent,
+  CrossChainEventLog,
+} from "./models/cross-chain-event";
+import {
+  createModelPM4PY,
+  checkConformancePM4PY,
+} from "./pm4py-adapter/ccmodel-adapter";
+
+export interface IWebAppOptions {
+  port: number;
+  hostname: string;
+}
+import {
+  CrossChainModel,
+  CrossChainModelType,
+  CrossChainTransactionSchema,
+  AssetState,
+} from "./models/crosschain-model";
+import {
+  BesuV2TxReceipt,
+  EthereumTxReceipt,
+  FabricV2TxReceipt,
+} from "./models/transaction-receipt";
+import { millisecondsLatency } from "./models/utils";
+import { RunTransactionV1Exchange as RunTransactionV1ExchangeBesu } from "@hyperledger/cactus-plugin-ledger-connector-besu";
+import { RunTransactionV1Exchange as RunTransactionV1ExchangeEth } from "@hyperledger/cactus-plugin-ledger-connector-ethereum";
+import { RunTxReqWithTxId } from "@hyperledger/cactus-plugin-ledger-connector-fabric";
+
+import { Observable } from "rxjs";
+import { filter, tap } from "rxjs/operators";
+
+export interface IPluginCcModelHephaestusOptions extends ICactusPluginOptions {
+  connectorRegistry?: PluginRegistry;
+  logLevel?: LogLevelDesc;
+  webAppOptions?: IWebAppOptions;
+  instanceId: string;
+  ethTxObservable?: Observable<RunTransactionV1ExchangeEth>;
+  besuTxObservable?: Observable<RunTransactionV1ExchangeBesu>;
+  fabricTxObservable?: Observable<RunTxReqWithTxId>;
+  sourceLedger: LedgerType;
+  targetLedger: LedgerType;
+}
+
+export class CcModelHephaestus implements ICactusPlugin, IPluginWebService {
+  private readonly log: Logger;
+  private readonly instanceId: string;
+  private endpoints: IWebServiceEndpoint[] | undefined;
+  private httpServer: Server | SecureServer | null = null;
+  private crossChainLog: CrossChainEventLog;
+  private unmodeledEventLog: CrossChainEventLog;
+  private nonConformedCrossChainLog: CrossChainEventLog;
+  private nonConformedCCTxs: string[];
```

Comment:
@RafaelAPB If these are not getting re-assigned anywhere outside of the constructor then you can also mark them `readonly` which doesn't change anything at runtime but it's a good practice to have as much of the state of your class being readonly as possible which helps lower bugs in this regard.

Reply:
okay, then the 3 Observables and `sourceLedger` and `targetLedger` should also be `readonly`, right?